Frameworks/CoreGraphics/DWriteFontCollectionHelper.mm

```diff
@@ -0,0 +1,207 @@
+    CFMutableArrayRef ret = CFArrayCreateMutable(kCFAllocatorSystemDefault, 0, &kCFTypeArrayCallBacks);
+
+    ComPtr<IDWriteFontList> fontList;
+    if (FAILED(_GetFontListForFamilyName(familyName, &fontList))) {
+        return ret;
+    }
+
+    size_t fontCount = fontList->GetFontCount();
+    for (size_t i = 0; i < fontCount; i++) {
+        ComPtr<IDWriteFont> font;
+        if (FAILED(fontList->GetFont(i, &font))) {
+            continue;
+        }
+
+        ComPtr<IDWriteLocalizedStrings> fullName;
+        BOOL exist = FALSE;
+        if (FAILED(font->GetInformationalStrings(DWRITE_INFORMATIONAL_STRING_FULL_NAME, &fullName, &exist))) {
+            continue;
+        }
+
+        if (exist) {
+            CFStringRef name = _CFStringFromLocalizedString(fullName.Get());
+            if (CFStringGetLength(name) == 0) {
+                TraceError(TAG, L"Failed to convert the localized string to wide string.");
+                continue;
+            }
+
+            CFArrayAppendValue(ret, name);
+        }
+    }
+
+    return ret;
+}
+
+std::shared_ptr<_DWriteFontProperties> DWriteFontCollectionHelper::GetFontPropertiesFromUppercaseFontName(
+    const woc::unique_cf<CFStringRef>& upperFontName) {
+    if (!m_propertiesMap) {
+        _InitializePropertiesMap();
+    }
+
+    const auto& info = m_propertiesMap->find(upperFontName);
+    if (info != m_propertiesMap->end()) {
+        return info->second;
+    }
+
+    return nullptr;
+}
+
+HRESULT DWriteFontCollectionHelper::CreateFontFamilyWithName(_In_ const wchar_t* unicharFamilyName,
+                                                             _Outptr_ IDWriteFontFamily** outFontFamily) {
+    ComPtr<IDWriteFontCollection> fontCollection = _GetFontCollection();
+
+    // If no/empty font collection, say that the font family couldn't be found
+    // Use S_FALSE to represent 'not found'
+    RETURN_HR_IF_NULL(S_FALSE, fontCollection);
+
+    size_t fontFamilyIndex;
+    BOOL fontFamilyExists;
+
+    RETURN_IF_FAILED(fontCollection->FindFamilyName(unicharFamilyName, &fontFamilyIndex, &fontFamilyExists));
+    if (fontFamilyExists) {
+        return fontCollection->GetFontFamily(fontFamilyIndex, outFontFamily);
+    }
+
+    return S_FALSE;
+}
+
+HRESULT DWriteFontCollectionHelper::_GetFontListForFamilyName(CFStringRef familyName, _Outptr_ IDWriteFontList** outFontList) {
+    ComPtr<IDWriteFontCollection> fontCollection = _GetFontCollection();
+
+    // If no/empty font collection, fail to return a font list
+    // Use of E_FAIL rather than a more nuanced HRESULT is permissable due to this being a private function only used in
+    // _InitializePropertiesMap()
+    RETURN_HR_IF_NULL(E_FAIL, fontCollection);
+
+    ComPtr<IDWriteFactory> dwriteFactory;
+    RETURN_IF_FAILED(DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED, __uuidof(IDWriteFactory), &dwriteFactory));
+
+    size_t index = 0;
+    BOOL exists = FALSE;
```

Comment:
>o>

---
In reply to: [90937198](https://github.com/Microsoft/WinObjC/pull/1504#discussion_r90937198) [](ancestors = 90937198)